internal/tasks/account_deletion.go

```diff
@@ -0,0 +1,213 @@
+	if err != nil {
+		return err
+	}
+	if errors.Is(err, sql.ErrNoRows) {
+		// assume the account got already deleted
+		return nil
+	}
+
+	actx := keppel.AuditContext{
+		UserIdentity: janitorUserIdentity{TaskName: "account-deletion"},
+		Request:      janitorDummyRequest,
+	}
+	err = j.DeleteAccount(ctx, *accountModel, actx)
+	if err != nil {
+		return fmt.Errorf("could not delete account %q: %w", accountName, err)
+	}
+	return nil
+}
+
+var (
+	deleteAccountFindManifestsQuery = sqlext.SimplifyWhitespace(`
+		SELECT r.name, m.digest
+			FROM manifests m
+			JOIN repos r ON m.repo_id = r.id
+			JOIN accounts a ON a.name = r.account_name
+			LEFT OUTER JOIN manifest_manifest_refs mmr ON mmr.repo_id = r.id AND m.digest = mmr.child_digest
+    WHERE a.name = $1 AND parent_digest IS NULL
+	`)
+	deleteAccountReposQuery                   = `DELETE FROM repos WHERE account_name = $1`
+	deleteAccountCountBlobsQuery              = `SELECT COUNT(id) FROM blobs WHERE account_name = $1`
+	deleteAccountScheduleBlobSweepQuery       = `UPDATE accounts SET next_blob_sweep_at = $2 WHERE name = $1`
+	deleteAccountMarkAllBlobsForDeletionQuery = `UPDATE blobs SET can_be_deleted_at = $2 WHERE account_name = $1`
+)
+
+type deleteAccountRemainingManifest struct {
+	RepositoryName string
+	Digest         string
+}
+
+func (j *Janitor) DeleteAccount(ctx context.Context, account models.Account, actx keppel.AuditContext) error {
+	// can only delete account when user has deleted all manifests from it
+	var nextManifests []deleteAccountRemainingManifest
+	err := sqlext.ForeachRow(j.db, deleteAccountFindManifestsQuery, []any{account.Name},
+		func(rows *sql.Rows) error {
+			var m deleteAccountRemainingManifest
+			err := rows.Scan(&m.RepositoryName, &m.Digest)
+			nextManifests = append(nextManifests, m)
+			return err
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	if len(nextManifests) > 0 {
+		// deletion was not completed yet -> check if we need to do something on our side
+		logg.Info("cleaning up account %q: need to delete %d manifests in this cycle", account.Name, len(nextManifests))
+		for _, rm := range nextManifests {
+			parsedDigest, err := digest.Parse(rm.Digest)
+			if err != nil {
+				return fmt.Errorf("while deleting manifest %q in repository %q: could not parse digest: %w",
+					rm.Digest, rm.RepositoryName, err)
+			}
+			repo, err := keppel.FindRepository(j.db, rm.RepositoryName, account.Name)
+			if err != nil {
+				return fmt.Errorf("while deleting manifest %q in repository %q: could not find repository in DB: %w",
+					rm.Digest, rm.RepositoryName, err)
+			}
+			err = j.processor().DeleteManifest(ctx, account.Reduced(), *repo, parsedDigest, actx)
+			if err != nil {
+				return fmt.Errorf("while deleting manifest %q in repository %q: %w",
+					rm.Digest, rm.RepositoryName, err)
+			}
+		}
+	}
+
+	// delete all repos (and therefore, all blob mounts), so that blob sweeping can immediately take place
```

Comment:
After deleting manifests found by `deleteAccountFindManifestsQuery`, you move directly into deleting repos. But at this point, it's not a given that you actually deleted all manifests. The query only deletes manifests not referenced by a parent manifest. To actually get rid of all manifests, you need to check `SELECT COUNT(*)` of the remaining manifests in the account, and restart from the top if there's still stuff left.

This does not come up in the tests because we don't have parent-child relationships there (like in multi-arch images). To exercise this, the managed account deletion test could be expanded by also pushing a list manifest that references the existing image manifest.

Reply:
nvm I found it